### PR TITLE
Improve energy efficiency by applying Dynamic Retry Delay Energy Pattern + Cache Energy Pattern 

### PR DESCRIPTION
**Pull request by anasofiagribeiro**
This improves the energy efficiency of SecondScreen by applying the Dynamic Retry Delay Energy Pattern and the Cache Energy Pattern for mobile applications.

The energy pattern was applied in RouterActivity.java and AndroidManifest.xml (for the Dynamic Retry Delay Energy Pattern) and TLSSocketFactoryCompat.java (for the Cache Energy Pattern). The general idea for the first one is to avoid accessing an internet connection if the connection is down . In particular, a class NetworkStateReceiver will be created in the java file that checks for an internet connection. The general idea for the second one is to avoid performing unnecessary operations by using cache mechanisms. In particular, setting the cache size to have no limit.

### CODE REVIEW

**Comment by wb9688** [comment]
Please follow our code style, so that Checkstyle won't complain.

**Comment by Redirion** [comment]
Sorry but this pattern appears to be outdated since Android 7, as the CONNECTIVITY_CHANGE cannot be received anymore from BroadcastReceivers registered in AndroidManifest.

Please see the Android doc: https://developer.android.com/reference/android/net/ConnectivityManager#CONNECTIVITY_ACTION

**Comment by anasofiagribeiro** [comment]
Hello. I have a question, if Android version 7 has API Level 24 (https://source.android.com/setup/start/build-numbers), how come "registerDefaultNetworkCallback", that was added in API Level 24, doesn't work? Thank you for your time.

**Comment by anasofiagribeiro** [comment]
Hello. Since the solution I proposed only works for Android 7 or newer, would a solution where this change only applies on Android 7 or newer be acceptable? This guarantees that for Android 7 or older it doesn’t break. 
“if (Build.VERSION.SDK_INT < 7 || hasActiveNetwork() ) {”

**Comment by Stypox** [comment]
> would a solution where this change only applies on Android 7 or newer be acceptable?

Yes, that's the way to go, thank you ;-)

File: app/src/main/java/org/schabi/newpipe/RouterActivity.java [hunk 1]
```diff
@@ -626,16 +631,26 @@ public void onCreate() {
 
         @Override
         protected void onHandleIntent(@Nullable final Intent intent) {
-            if (intent == null) {
-                return;
-            }
+            if (Build.VERSION.SDK_INT < 7 || hasActiveNetwork()) {
```
Comment (Redirion) [inline app/src/main/java/org/schabi/newpipe/RouterActivity.java:634]:
```suggestion
            if (Build.VERSION.SDK_INT < 24 || hasActiveNetwork()) {
```

File: app/src/main/java/org/schabi/newpipe/RouterActivity.java [hunk 2]
```diff
@@ -745,6 +760,71 @@ public void onDestroy() {
...
+                    final AlarmManager alarmManager = (AlarmManager)
+                            context.getSystemService(Context.ALARM_SERVICE);
+                    final Intent innerIntent = new Intent(context, FetcherService.class);
+                    final PendingIntent pendingIntent =
+                            PendingIntent.getService(context, 0, innerIntent, 0);
+
+                    SharedPreferences preferences =
+                            context.getSharedPreferences(context.getPackageName(),
+                                    Context.MODE_PRIVATE);
+                    preferences.edit();
```
Comment (Stypox) [inline app/src/main/java/org/schabi/newpipe/RouterActivity.java:796]: Why `.edit()`?

**Review by Stypox — CHANGES_REQUESTED** [review]
It doesn't build because of checkstyle, see https://travis-ci.org/github/TeamNewPipe/NewPipe/builds/716366001
Also, could you further explain what you did? Is it safe to put the ssl cache to infinity? And are we sure we want to do nothing silently when the user shares something to NewPipe until the connection is active (if I understood correctly that's what the code does)?

File: app/src/main/java/org/schabi/newpipe/RouterActivity.java (hunk 2 above)
Comment (Redirion) [reply app/src/main/java/org/schabi/newpipe/RouterActivity.java:796]:
```suggestion
```
unintente obviosly

**Comment by mhmdanas** [comment]
@anasofiagribeiro  I'm closing this PR, but please make a comment when you want to resume work on this so I would re-open it. Thanks!

**Pull request closed without merging**